### PR TITLE
Add symfony events for login action

### PR DESCRIPTION
**Pull request by sharidas**
Add symfony events for login action.
The symfony events are added for:
1 - login with token
2 - login with apache

Signed-off-by: Sujith H <sharidasan@owncloud.com>

## Description

Adding symfony events for the missing login methods.

## Related Issue

https://github.com/owncloud/core/issues/31956

## Motivation and Context

Adding symfony events for the missing login methods.

## How Has This Been Tested?

[x] Tested by adding unit tests.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by sharidas** [comment]
The reason for using `login` instead of `uid` is because of https://github.com/owncloud/core/blob/master/lib/private/User/Session.php#L505.

**Review by PVince81 — COMMENTED** [review]
This looks good.

I wonder if we should use the opportunity to add another argument "$loginType" that tells whether login was done with password, with token or apache in case it would be useful for apps.

**Comment by sharidas** [comment]
> I wonder if we should use the opportunity to add another argument "$loginType" that tells whether login was done with password, with token or apache in case it would be useful for apps.

For the event listeners it would be an added advantage. I will add the change to this PR. Thanks.

**Comment by sharidas** [comment]
Added :
`'loginType' => 'password'`, `'loginType' => 'token'` and `'loginType' => 'apache'`
Also updated the tests so that mocked EventDispatcher is used, to improve readability of the test code.

File: tests/lib/User/SessionTest.php [hunk 1]
```diff
@@ -1540,6 +1586,15 @@ function (GenericEvent $event) use (&$calledFailedLoginEvent) {
 			->method('getUID')
 			->willReturn('foo');
 
+		$failedEvent = new GenericEvent(null, ['user' => 'foo']);
+		//$beforeloginEvent = new GenericEvent(null, ['user' => $iUser, 'uid' => 'foo', 'password' => 'foo']);
```
Comment (PVince81) [inline tests/lib/User/SessionTest.php]: remove?

File: tests/lib/User/SessionTest.php [hunk 2]
```diff
@@ -1540,6 +1586,15 @@ function (GenericEvent $event) use (&$calledFailedLoginEvent) {
...
+		$eventDispatcher->expects($this->once())
+			->method('dispatch')
+			->withConsecutive(
+				//[$this->equalTo('user.beforelogin'), $this->equalTo($beforeloginEvent)],
```
Comment (PVince81) [inline tests/lib/User/SessionTest.php]: remove ? or is this not covered yet ?

**Review by PVince81 — COMMENTED** [review]
Thanks for the addition.

Please remove or integrate the commented out test code

**Comment by PVince81** [comment]
@butonic @tomneedham FYI https://github.com/owncloud/core/pull/31963#issuecomment-402444259

File: tests/lib/User/SessionTest.php (hunk 2 above)
Comment (sharidas) [reply tests/lib/User/SessionTest.php]: Yes need to remove this line.

File: tests/lib/User/SessionTest.php (hunk 1 above)
Comment (sharidas) [reply tests/lib/User/SessionTest.php]: Yes need to remove this. Removed in the updated version.

**Review by PVince81 — APPROVED** [review]
:+1:

**Comment by sharidas** [comment]
Backport PR: https://github.com/owncloud/core/pull/31985

**Pull request merged**